File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character,Character>map=new HashMap<>();
        Stack<Character>st=new Stack<>();
        map.put(')','(');
        map.put('}','{');
        map.put(']','[');
        for(char c:s.toCharArray()){
            if(map.containsKey(c)){
                char top=st.isEmpty()?'#':st.pop();
                if(top!=map.get(c)){
                    return false;
                }
            }
            else{
                st.push(c);
            }
        }
        return st.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for ch in s:
            if ch in "({[":
                stack.append(ch)
            else:
                if not stack:
                    return False

                top = stack.pop()

                if ((ch == ')' and top != '(') or
                    (ch == '}' and top != '{') or
                    (ch == ']' and top != '[')):
                    return False

        return not stack